lib/DriverTool/swift_parse_test_main.cpp

```diff
@@ -0,0 +1,235 @@
+#include "swift/Basic/LLVM.h"
+#include "swift/Basic/LangOptions.h"
+#include "swift/Bridging/ASTGen.h"
+#include "swift/Parse/Parser.h"
+#include "llvm/Support/CommandLine.h"
+
+#include <chrono>
+#include <ctime>
+
+using namespace swift;
+
+namespace {
+
+enum class Executor {
+  SwiftParser,
+  LibParse,
+};
+
+enum class ExecuteOptionFlag {
+  /// Enable body skipping
+  SkipBodies = 1 << 0,
+};
+using ExecuteOptions = OptionSet<ExecuteOptionFlag>;
+
+struct SwiftParseTestOptions {
+  llvm::cl::list<Executor> Executors = llvm::cl::list<Executor>(
+      llvm::cl::desc("Available parsers"),
+      llvm::cl::values(
+          clEnumValN(Executor::SwiftParser, "swift-parser", "SwiftParser"),
+          clEnumValN(Executor::LibParse, "lib-parse", "libParse")));
+
+  llvm::cl::opt<unsigned int> Iteration = llvm::cl::opt<unsigned int>(
+      "n", llvm::cl::desc("iteration"), llvm::cl::init(1));
+
+  llvm::cl::opt<bool> SkipBodies = llvm::cl::opt<bool>(
+      "skip-bodies",
+      llvm::cl::desc("skip function bodies and type members if possible"));
+
+  llvm::cl::list<std::string> InputPaths = llvm::cl::list<std::string>(
+      llvm::cl::Positional, llvm::cl::desc("input paths"));
+};
+
+struct LibParseExecutor {
+  constexpr static StringRef name = "libParse";
+
+  static void performParse(llvm::MemoryBufferRef buffer, ExecuteOptions opts) {
+    SourceManager SM;
+    unsigned bufferID =
+        SM.addNewSourceBuffer(llvm::MemoryBuffer::getMemBuffer(buffer));
+    DiagnosticEngine diagEngine(SM);
+    LangOptions langOpts;
+    TypeCheckerOptions typeckOpts;
+    SILOptions silOpts;
+    SearchPathOptions searchPathOpts;
+    ClangImporterOptions clangOpts;
+    symbolgraphgen::SymbolGraphOptions symbolOpts;
+    std::unique_ptr<ASTContext> ctx(
+        ASTContext::get(langOpts, typeckOpts, silOpts, searchPathOpts,
+                        clangOpts, symbolOpts, SM, diagEngine));
+
+    SourceFile::ParsingOptions parseOpts;
+    parseOpts |= SourceFile::ParsingFlags::DisablePoundIfEvaluation;
+    if (!opts.contains(ExecuteOptionFlag::SkipBodies))
+      parseOpts |= SourceFile::ParsingFlags::DisableDelayedBodies;
+
+    ModuleDecl *M = ModuleDecl::create(Identifier(), *ctx);
+    SourceFile *SF =
+        new (*ctx) SourceFile(*M, SourceFileKind::Library, bufferID, parseOpts);
+
+    Parser parser(bufferID, *SF, /*SILParserState=*/nullptr);
+    SmallVector<ASTNode> items;
+    parser.parseTopLevelItems(items);
+  }
+};
+
+struct SwiftParserExecutor {
+  constexpr static StringRef name = "SwiftParser";
+
+  static void performParse(llvm::MemoryBufferRef buffer, ExecuteOptions opts) {
+#if SWIFT_BUILD_SWIFT_SYNTAX
+    // TODO: Implement 'ExecuteOptionFlag::SkipBodies'
+    auto sourceFile = swift_ASTGen_parseSourceFile(
+        buffer.getBufferStart(), buffer.getBufferSize(), /*moduleName=*/"",
+        buffer.getBufferIdentifier().data(), /*ASTContext=*/nullptr);
+    swift_ASTGen_destroySourceFile(sourceFile);
+#endif
```

Comment:
Doesn’t really matter since it’s a test utility but should we error if we don’t have `SWIFT_BUILD_SWIFT_SYNTAX`?

Reply:
We do in `swift_parse_test_main`, I also made this mistake when I first went through it. An alternative design would be to have these methods return an `Error`, which would mean the SWIFT_BUILD_SWIFT_SYNTAX check in `swift_parse_test_main` could be removed for something more generic.